packages/data/addon/utils/classes/cache.ts

```diff
@@ -0,0 +1,124 @@
+/**
+ * Copyright 2022, Yahoo Holdings Inc.
+ * Licensed under the terms of the MIT license. See accompanying LICENSE.md file for terms.
+ */
+
+type CacheRecord<T> = {
+  creationTime: number; // the time this record was created
+  lastAccessed: number; // the last time this record was accessed
+  records: T; // the actual records
+};
+
+/**
+ * a cache
+ * to use, give it a unique string (cacheKey) and whatever you want stored alongside that key
+ * returns only exact matches
+ */
+export default class Cache<T> {
+  private _maxCacheSize;
+  private _cacheTimeout;
+  private _cache;
+  getCacheKey;
+
+  /**
+   * @param {function (params:type) => string} cacheKeyFunction - a function that returns unique strings associated with specific items in cache
+   * @param {number} maxCacheSize - the maximum number of items that can be stored in the cache (default 10)
+   * @param {number} cacheTimeout - the maximum amount of milliseconds a record should be allowed to stay in the cache (default 1 hour)
+   */
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  constructor(cacheKeyFunction: (args: any) => string, maxCacheSize = 10, cacheTimeout = 60 * 60 * 1000) {
+    this.getCacheKey = cacheKeyFunction;
+    this._maxCacheSize = maxCacheSize;
+    this._cacheTimeout = cacheTimeout;
+    this._cache = new Map<string, CacheRecord<T>>();
+  }
+
+  /**
+   * removes items that were created too long ago from the cache
+   */
+  private _removeStaleItems(): void {
+    const cache = this._cache;
+    const now = Date.now();
+    cache.forEach((value: CacheRecord<T>, key: string) => {
+      if (now - value.creationTime >= this._cacheTimeout) {
+        cache.delete(key);
+      }
+    });
+  }
+
+  /**
+   * removes the least recently accessed items until cache is at or below max size
+   * @returns {number} the number of items removed from the cache
+   */
+  private _trimExtraItems(): number {
+    const cache = this._cache;
+    let removedCount = 0;
+    while (cache.size > this._maxCacheSize) {
+      let oldestDate = Date.now();
+      let oldestKey = '';
+      cache.forEach((value: CacheRecord<T>, key: string) => {
+        if (value.lastAccessed < oldestDate) {
+          oldestDate = value.lastAccessed;
+          oldestKey = key;
+        }
+      });
+      cache.delete(oldestKey);
+
+      removedCount = removedCount + 1;
+    }
+    return removedCount;
+  }
```

Comment:
Since the interface only allows you to set one item at a time, I think we should be able to make something simpler like

```js
_makeSpaceFor(cacheKey) {
    const cache = this._cache;
    // if the key is not in the cache and there is no space left
    if(!cache.has(cacheKey) && cache.size === this._maxCacheSize) {
        // find and remove least recently accessed
   }
}

setItem(cacheKey, records) {
    this._removeStaleItems();
    if (!cacheKey) {
      return;
    }
	this._makeSpaceFor(cacheKey);
    //....
}
```